Separate operator generation and building logic in scripts

mangelajo opened the pull request:
The scripts are: codegen-operator and build-operator, we rely
on a stable dapper based environment to be able to generate
and build across the team without issues.

The dapper image now installs operator-sdk from an specific
master commit it's know for us to work.

To regenerate the operators/go/submariner-operator:

    make codegen-operator

To build the operator container image:

    make build-operator

This PR contains a second commit which will use the build-operator
images when testing e2e, instead of using the ones from quay.

mangelajo committed:
Separate operator generation and building logic in scripts

The scripts are: codegen-operator and build-operator, we rely
on a stable dapper based environment to be able to generate
and build across the team without issues.

The dapper image now installs operator-sdk from an specific
master commit it's know for us to work.

To regenerate the operators/go/submariner-operator:

    make codegen-operator

To build the operator container image:

    make build-operator

mangelajo commented:
I still think we are at the right moment to rip just commit the operator scaffold source to the repo

dfarrell07 commented on operators/go/gen_subm_operator.sh:
```diff
@@ -154,10 +154,9 @@ function build_subm_operator() {
   pushd $op_dir
   go mod vendor
   # This seems like a bug in operator-sdk, that this is needed?
-  # NB: These fail for some people with missing "latest" tag error
```
Maybe this can be a different PR? I think the discrepancy different folks are seeing with these prereqs deserves more digging, but I'd love to go ahead and merge the rest of this.

mangelajo replied:
I have separated the logic, and if you try the make codegen-operator && make build-operator you will see those are not necessary anymore with the current operator-sdk

dfarrell07 commented on operators/go/gen_subm_operator.sh:
```diff
@@ -154,10 +154,9 @@ function build_subm_operator() {
-  go get k8s.io/kube-state-metrics/pkg/collector || true
-  go get k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1 || true
-  go get github.com/coreos/prometheus-operator/pkg/apis/monitoring ||
+  # go get k8s.io/kube-state-metrics/pkg/collector
```
This causes a failure for me:

```
+ operator-sdk build quay.io/submariner/submariner-operator:0.0.1 --verbose
DEBU[0000] Debug logging is set                         
DEBU[0000] Running []string{"go", "build", "-o", "/home/daniel/go/src/github.com/submariner-operator/submariner-operator/build/_output/bin/submariner-operator", "-gcflags", "all=-trimpath=/home/daniel/go/src/github.com/submariner-operator", "-asmflags", "all=-trimpath=/home/daniel/go/src/github.com/submariner-operator", "-mod=vendor", "github.com/submariner-operator/submariner-operator/cmd/manager"} 
build k8s.io/kube-state-metrics/pkg/collector: cannot find module for path k8s.io/kube-state-metrics/pkg/collector
Error: failed to build operator binary: (failed to exec []string{"go", "build", "-o", "/home/daniel/go/src/github.com/submariner-operator/submariner-operator/build/_output/bin/submariner-operator", "-gcflags", "all=-trimpath=/home/daniel/go/src/github.com/submariner-operator", "-asmflags", "all=-trimpath=/home/daniel/go/src/github.com/submariner-operator", "-mod=vendor", "github.com/submariner-operator/submariner-operator/cmd/manager"}: exit status 1)
```

mangelajo replied:
Let's make sure it works from dapper, and then we can all use dapper and stop relying on inconsistent environments, and for it to work in dapper we need to remove that (at least with the operator-sdk built from master, I'm trying now to make "go get" work)

mangelajo replied:
If you try from dapper you will get operator-sdk from master, in that case not even the other go-gets are necessary.

mkolesnik commented:
I agree, no reason to regenerate each time since it's more trouble than it's worth IMO.

The gen script is still useful as an automated way to generate the operator entirely, but the scaffold can be thrown away and we can use the existing controllers instead.

mangelajo commented:
@dfarrell07 I seem to have no write access to this repo, can you check?,

github says: "Only those with write access to this repository can merge pull requests."

mkolesnik commented on scripts/codegen-operator:
```diff
@@ -0,0 +1,32 @@
+kind delete cluster || true
+kind create cluster
+
+git clone https://github.com/operator-framework/operator-sdk
```
Why don't you use go get?
`go get -d github.com/operator-framework/operator-sdk`

mangelajo replied:
did that work for you?, it didn't for me, but let me try with your exact command.

This is used by make codegen-operator but I will update the README.md, you're right

mangelajo replied:
I tried that, but afterwards can't find operator-sdk installed anywhere (path or not...) 
```
root@cd8819185765:/go/src/github.com/submariner-io/submariner# find / -name "operator-sdk"
/go/pkg/mod/github.com/operator-framework/operator-sdk@v0.10.0/cmd/operator-sdk
/go/pkg/mod/cache/download/github.com/operator-framework/operator-sdk
/go/src/github.com/submariner-io/submariner/operators/go/submariner-operator/vendor/github.com/operator-framework/operator-sdk
root@cd8819185765:/go/src/github.com/submariner-io/submariner# /go/pkg/mod/github.com/operator-framework/operator-sdk@v0.10.0/cmd/operator-sdk
bash: /go/pkg/mod/github.com/operator-framework/operator-sdk@v0.10.0/cmd/operator-sdk: Is a directory
```

mangelajo replied:
outdated now

mkolesnik commented on scripts/codegen-operator:
```diff
@@ -0,0 +1,32 @@
+mkdir -p $OP_FRAMEWORK
+pushd $OP_FRAMEWORK
+
+kind delete cluster || true
```
Why is a kind cluster needed here? This logic is already in the gen script..

mangelajo replied:
To compiler operator-sdk it's also necessary... not to install from binary, but binary does not work.

mkolesnik commented on scripts/codegen-operator:
```diff
@@ -0,0 +1,32 @@
+
+popd
+cd operators
+cd go
```
cd operators/go ?

mangelajo replied:
:-)

mkolesnik commented on scripts/codegen-operator:
```diff
@@ -0,0 +1,32 @@
+cd operators
+cd go
+./gen_subm_operator.sh
+
```
nit: unnecessary new line

mkolesnik reviewed (changes requested):
I don't understand how this script is used, are you calling it from the e2e? I don't see you're added any call nor documentation on how to use it.

dfarrell07 reviewed (commented):
> @dfarrell07 I seem to have no write access to this repo, can you check?,
> 
> github says: "Only those with write access to this repository can merge pull requests."

GitHub says you need to accept your invite to be a Contributor.

mangelajo commented:
Give it a try, you can run:

make codegen-operator

then

make build-operator

mangelajo committed:
Make e2e work without external dependencies

Builds on the previous codegen-operator and build-operator patch,
and pushes the submariner image to the local kind clusters before
deployment.

Adds a custom deployment yaml file that sources the right image.

mkolesnik commented on scripts/kind-e2e/e2e.sh:
```diff
@@ -3,6 +3,8 @@ set -em
 
 source $(git rev-parse --show-toplevel)/scripts/lib/debug_functions
 
+deploytool=$5
```
I already introduced a variable `deploy_operator` so you can use that, e.g.
`if [ "$deploy_operator" = true ]; then`

mkolesnik commented on scripts/e2e:
```diff
@@ -5,6 +5,11 @@ source $(dirname $0)/lib/debug_functions
 
+if [[ $5 = operator ]]; then
+   test -d ../operators/go/submariner-operator || ./codegen-operator
+   # ./build-operator
```
Delete this?

mangelajo replied:
True

mangelajo replied:
Oh no, I should un-comment, I skipped the build to test e2e once and again without waiting for the operator to be re-built.

mkolesnik commented on scripts/kind-e2e/lib_operator_deploy_subm.sh:
```diff
@@ -5,30 +5,10 @@ if [ "${0##*/}" = "lib_operator_deploy_subm.sh" ]; then
-cp -a $subm_op_src_dir/. $subm_op_dir/
-
-export GO111MODULE=on
+subm_ns=operators
```
This is already set in `create_subm_vars`

mkolesnik commented on scripts/kind-e2e/lib_operator_deploy_subm.sh:
```diff
@@ -5,30 +5,10 @@ if [ "${0##*/}" = "lib_operator_deploy_subm.sh" ]; then
-
-export GO111MODULE=on
+subm_ns=operators
+subm_broker_ns=submariner-k8s-broker
```
ditto

mkolesnik commented on Dockerfile.dapper:
```diff
@@ -45,6 +45,12 @@ RUN apt-get -q update && \
     GOFLAGS="" go get -v github.com/onsi/ginkgo/ginkgo && \
     GOFLAGS="" go get -v golang.org/x/tools/cmd/goimports
 
+RUN OP_FRAMEWORK="$GOPATH/src/github.com/operator-framework" && \
```
Still hacky :P

We should make it more robust in the future, maybe stick a TODO here?

mangelajo replied:
ack, makes sense, I don't know why the guide instructions don't work here.

mangelajo replied:
Done

mkolesnik commented on operators/go/build_subm_operator.sh:
```diff
@@ -0,0 +1,25 @@
+op_gen_dir=$(pwd)
+op_out_dir=$op_gen_dir/submariner-operator
+
+function build_subm_operator() {
```
You can just drop this and run the commands :)

mangelajo replied:
true

mangelajo replied:
done

mkolesnik commented on scripts/build-operator:
```diff
@@ -0,0 +1,10 @@
+
+source $(dirname $0)/lib/debug_functions
+
+cd $(dirname $0)
```
w00t?

mangelajo replied:
Yes, because you can be calling this script from other directories.

I will squash this to the other "cd" down L9

mkolesnik commented on scripts/codegen-operator:
```diff
@@ -0,0 +1,10 @@
+
+source $(dirname $0)/lib/debug_functions
+
+cd $(dirname $0)
```
w00t?

mangelajo replied:
ditto :)

mkolesnik commented on scripts/build-operator:
```diff
@@ -0,0 +1,10 @@
+
+cd $(dirname $0)
+
+GOFLAGS=""
```
This necessary? If so, shouldn't it be exported?

mangelajo replied:
hmmm good question, since it's not exported I suspect it's not?

mkolesnik commented on scripts/codegen-operator:
```diff
@@ -0,0 +1,10 @@
+
+cd $(dirname $0)
+
+GOFLAGS=""
```
This necessary? If so, shouldn't it be exported?

mkolesnik reviewed (changes requested):
Looks good, some small changes

mangelajo committed:
Address reviewer comments

dfarrell07 commented on Dockerfile.dapper:
```diff
@@ -32,7 +32,7 @@ ENV GOLANG_ARCH_amd64=amd64 GOLANG_ARCH_arm=armv6l GOLANG_ARCH=GOLANG_ARCH_${ARC
 # goimports      | code formatting
 RUN apt-get -q update && \
-    apt-get install -y gcc git curl docker.io && \
+    apt-get install -y gcc git curl docker.io mercurial make && \
```
Mercurial seems to take a really long time to install FYI. My build hangs on it, with one thread at 100%, for what feels like minutes.

dfarrell07 commented:
This works great for me btw, thanks! :+1:

dfarrell07 commented:
Using it more, I think there are two issues (with the build/push logic) this change introduces:

Pushing Operator images in new containerized build env fails due to lack of auth: https://jira.coreos.com/browse/HYCLD-268

Fix/doc Operator version/push_image vars in new containerized build env: https://jira.coreos.com/browse/HYCLD-269

But otherwise it's still wonderful, thanks again @mangelajo! :pray:
